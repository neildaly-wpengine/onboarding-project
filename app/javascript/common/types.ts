export interface Article {
    title: string;
    content: string;
    createdAt: string;
    updatedAt: string;
    discardedAt?: string;
    archived: boolean;
    id: string;
    user: User;
    // todo remove this from backend serializer
    userID: number;
}

export interface User {
    id: string;
    firstName: string;
    lastName: string;
    initialsImageLink: string;
}