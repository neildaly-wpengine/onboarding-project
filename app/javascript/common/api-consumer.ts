import axios, { AxiosResponse } from "axios";
import humps from 'humps';
import {
    Article,
    CreateArticleBody,
    LoginBody,
    LogoutResponse,
    Registration,
    RegistrationBody,
    DeleteResponse
} from "./types";
const JSONAPIDeserializer = require("jsonapi-serializer").Deserializer;

class APIConsumer {
    async getAllArticles(): Promise<Article[]> {
        try {
            const response: AxiosResponse = await axios.get("/api/v1/articles");
            const articleData: Article[] = await this.deserializeResponse(response);

            return articleData;
        } catch (e) {
            throw e
        }
    }

    async getSpecificArticle(id: string): Promise<Article> {
        try {
            const response: AxiosResponse = await axios.get(`/api/v1/articles/${id}`);
            const articleData: Article = await this.deserializeResponse(response);

            return articleData;
        } catch (e) {
            throw e;
        }
    }

    async registerNewUser(registrationBody: RegistrationBody): Promise<any> {
        try {
            const decamelizedKeys: RegistrationBody = humps.decamelizeKeys(registrationBody) as RegistrationBody;
            const response: Registration = await axios.post('/api/v1/registrations', decamelizedKeys,
                { withCredentials: true });

            return humps.camelizeKeys(response);
        } catch (e) {
            throw e;
        }
    }

    async destroySession(): Promise<LogoutResponse> {
        try {
            const response: AxiosResponse = await axios.delete('/api/v1/logout');

            return humps.camelizeKeys(response) as LogoutResponse;
        } catch (e) {
            throw e;
        }
    }

    async loginUser(loginBody: LoginBody): Promise<any> {
        try {
            const response: AxiosResponse = await axios.post('/api/v1/sessions', loginBody,
                { withCredentials: true });

            return humps.camelizeKeys(response);
        } catch (e) {
            throw e;
        }
    }

    async createArticle(articleBody: CreateArticleBody): Promise<Article> {
        try {
            const response: AxiosResponse = await axios.post('/api/v1/articles',
                humps.decamelizeKeys(articleBody),
                { withCredentials: true });
            const articleData: Article = await this.deserializeResponse(response);

            return articleData;
        } catch (e) {
            throw e;
        }
    }

    async editArticle(articleBody: CreateArticleBody, articleID: string): Promise<Article> {
        try {
            const response: AxiosResponse = await axios.patch(`/api/v1/articles/${articleID}`,
                humps.decamelizeKeys(articleBody),
                { withCredentials: true });
            const articleData: Article = await this.deserializeResponse(response);

            return articleData;
        } catch (e) {
            throw e;
        }
    }

    async deleteArticle(articleID: string): Promise<DeleteResponse> {
        try {
            const response: AxiosResponse = await axios.delete(`/api/v1/articles/${articleID}`,
                { withCredentials: true });

            return humps.camelizeKeys(response) as DeleteResponse;
        } catch (e) {
            throw e;
        }
    }

    async recoverArticle(articleID: string): Promise<Article> {
        try {
            const response: AxiosResponse = await axios.post(`/api/v1/articles/${articleID}/recover`,
                { withCredentials: true });
            const articleData: Article = await this.deserializeResponse(response);

            return articleData;
        } catch (e) {
            throw e;
        }
    }

    private deserializeResponse = (response: AxiosResponse) => {
        const deserializerOptions = {
            keyForAttribute: "camelCase"
        }
        const deserializer = new JSONAPIDeserializer(deserializerOptions);
        return deserializer.deserialize(
            response.data,
            (_err: any, articlesResponseData: any) => {
                return articlesResponseData;
            }
        );
    }
}

export default APIConsumer;
