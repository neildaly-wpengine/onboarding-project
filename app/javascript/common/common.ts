import { User } from "./types";

export const createUserInitials = (user: User): string => {
    const userInitials: string = `${user.firstName.charAt(
        0
    )}${user.lastName.charAt(0)}`.toUpperCase();

    return userInitials;
}
