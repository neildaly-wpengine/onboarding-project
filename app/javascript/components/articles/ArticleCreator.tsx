import AddCircleIcon from "@material-ui/icons/AddCircle";
import React, { useState } from "react";
import { Redirect } from "react-router";
import {
  ArticleCreationContent,
  AuthStoreProps,
  ConsumerProps,
  CreateArticleBody,
} from "../../common/types";
import ArticleForm from "./ArticleForm";

const ArticleCreator: React.FC<ConsumerProps & AuthStoreProps> = ({
  consumer,
  authStore,
}) => {
  const [created, setCreated] = useState<boolean>(false);
  const [articleBody, setArticleBody] = useState<ArticleCreationContent>({
    title: "",
    content: "",
    userId: parseInt(authStore.user.id),
  });

  const handleChange = (e: React.BaseSyntheticEvent): void => {
    const { name, value } = e.target;
    setArticleBody({ ...articleBody, [name]: value });
  };

  const handleSubmit = async (e: React.BaseSyntheticEvent) => {
    e.preventDefault();

    const response = await consumer.createArticle({
      article: articleBody,
    } as CreateArticleBody);

    setCreated(() => {
      return Boolean(response.createdAt);
    });
  };

  if (created) {
    return <Redirect to={{ pathname: "/", state: { created: true } }} />;
  }

  if (!authStore.authenticated) {
    return <Redirect to="/" />;
  }

  return (
    <ArticleForm
      handleChange={handleChange}
      handleSubmit={handleSubmit}
      icon={<AddCircleIcon />}
      formTitle="Create Article"
      buttonText="Create"
    />
  );
};

export default ArticleCreator;
