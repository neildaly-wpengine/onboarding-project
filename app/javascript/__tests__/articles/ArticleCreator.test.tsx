import "@testing-library/jest-dom";
import "@testing-library/jest-dom/extend-expect";
import {
  act,
  cleanup,
  fireEvent,
  render,
  RenderResult,
} from "@testing-library/react";
import axios from "axios";
import humps from "humps";
import * as React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { AuthStore } from "../../common/types";
import ArticleCreator from "../../components/articles/ArticleCreator";
import {
  consumer,
  jsonAPISpecificArticleResponseData as response,
  mockUser,
} from "./__helpers__/test-data";

const renderArticleCreator = (authenticated = true): RenderResult => {
  const mockAuthStore: AuthStore = {
    authenticated: authenticated,
    user: mockUser,
  };
  return render(
    <Router>
      <ArticleCreator consumer={consumer} authStore={mockAuthStore} />
    </Router>
  );
};

afterEach(() => {
  cleanup;
  jest.resetAllMocks();
});

jest.mock("axios");

describe("<ArticleCreator />", () => {
  describe("Content", () => {
    it("renders correctly", () => {
      const { container } = renderArticleCreator();

      expect(container).toMatchSnapshot();
    });

    it("does not render when not authenticated", () => {
      const { container } = renderArticleCreator(false);

      expect(container).toMatchSnapshot();
      expect(container).toBeEmptyDOMElement();
    });

    it("successfully creates an article and redirects", async () => {
      const { getByTestId, container } = renderArticleCreator();
      const redirectURL: string = "/";
      const titleInput = getByTestId("titleInput");
      const contentInput = getByTestId("contentInput");
      const submit = getByTestId("submit");
      const expectedResponse = {
        data: {
          data: {
            response,
          },
        },
      };

      act(() => {
        fireEvent.change(titleInput, {
          target: {
            value: response.data.attributes.title,
          },
        });
      });
      act(() => {
        fireEvent.change(contentInput, {
          target: {
            value: response.data.attributes.content,
          },
        });
      });

      (axios.post as jest.Mock).mockImplementationOnce(() =>
        Promise.resolve(expectedResponse)
      );

      fireEvent.submit(submit);

      expect(axios.post).toHaveBeenCalledTimes(1);
      expect(axios.post).toHaveBeenCalledWith(
        "/api/v1/articles",
        humps.decamelizeKeys({
          article: {
            title: response.data.attributes.title,
            content: response.data.attributes.content,
            userId: parseInt(response.included[0].id),
          },
        }),
        { withCredentials: true }
      );
      expect(container.innerHTML).toEqual(expect.stringContaining(redirectURL));
    });
  });
});
