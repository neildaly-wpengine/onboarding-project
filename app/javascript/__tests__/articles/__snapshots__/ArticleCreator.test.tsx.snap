// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleCreator /> Content does not render when not authenticated 1`] = `<div />`;

exports[`<ArticleCreator /> Content renders correctly 1`] = `
<div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthSm"
  >
    <div
      class="makeStyles-paper-1"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-2 MuiAvatar-colorDefault"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </div>
      <h1
        class="MuiTypography-root MuiTypography-h5"
      >
        Create Article
      </h1>
      <form
        class="makeStyles-form-3"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused Mui-required Mui-required"
                data-shrink="true"
                for="title"
                id="title-label"
              >
                Title
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  data-testid="titleInput"
                  id="title"
                  name="title"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-6 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-8 PrivateNotchedOutline-legendNotched-9"
                  >
                    <span>
                      Title
                       *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="content"
                id="content-label"
              >
                Content
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                  data-testid="contentInput"
                  id="content"
                  name="content"
                  required=""
                  rows="1"
                  style="height: 0px; overflow: hidden;"
                />
                <textarea
                  aria-hidden="true"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                  readonly=""
                  style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
                  tabindex="-1"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-6 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-8"
                  >
                    <span>
                      Content
                       *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-4 MuiButton-containedPrimary MuiButton-fullWidth"
          data-testid="submit"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Create
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </main>
</div>
`;
